Mini rework and tests for Path utils

reichlfl opened the pull request:
This came up for a method that was added to Path.cpp in this PR: https://github.com/google/orbit/pull/1120
I decided to consistently add [[nodiscard]] for all methods, and this lead to some more fixes... which lead to tests to be sure I didn't break the rest of the class.

- Added [[nodiscard]] everywhere
- Changed internal logic for GetBasePath() and Init() due to a warning resulting from [[nodiscard]]
- Renamed some methods to make it more obvious that they are creating directories, and to distinguish between "path" and "dir"
- Added tests for uncovered parts of Path.cpp to make sure I didn't break anything

dimitry- commented on OrbitQt/orbitmainwindow.cpp:
```diff
@@ -189,7 +189,7 @@ OrbitMainWindow::OrbitMainWindow(QApplication* a_App, ApplicationOptions&& optio
 
   SetTitle({});
-  std::string iconFileName = Path::GetExecutablePath() + "orbit.ico";
+  std::string iconFileName = Path::GetExecutableDir() + "orbit.ico";
```
Path::Join

reichlfl replied:
Done

dimitry- commented on OrbitCore/Path.cpp:
```diff
@@ -97,13 +92,10 @@ bool Path::DirExists(const std::string& dir) {
 }
 
 std::string Path::GetBasePath() {
```
this is used only Dllpaht which is not used anywhere else - remove this and GetDllPath and base_path_

antonrohr replied:
this result then in Path not having any state anymore and only static functions. You can then consider changing it from a class to a namespace Path. (I have a slight preference towards namespace instead static class, but do not care to much)

reichlfl replied:
True, I changed it to a namespace now, thanks

dimitry- commented on OrbitCore/Path.h:
```diff
@@ -10,48 +10,47 @@
+  [[nodiscard]] static std::string GetExecutablePath();
+  [[nodiscard]] static std::string GetExecutableDir();
+  [[nodiscard]] static std::string GetBasePath();
+  [[nodiscard]] static std::string GetDllPath(bool a_Is64Bit);
```
remove this.

reichlfl replied:
Done

dimitry- commented on OrbitCore/Path.h:
```diff
@@ -10,48 +10,47 @@
+  [[nodiscard]] static std::string GetExecutableDir();
+  [[nodiscard]] static std::string GetBasePath();
+  [[nodiscard]] static std::string GetDllPath(bool a_Is64Bit);
+  [[nodiscard]] static std::string GetDllName(bool a_Is64Bit);
```
is_64bit

reichlfl replied:
GetDllName is removed anyways

dimitry- commented on OrbitCore/Path.h:
```diff
@@ -10,48 +10,47 @@
+  [[nodiscard]] static std::string GetSymbolsFileName();
+  [[nodiscard]] static std::string CreateAndGetCacheDir();
+  [[nodiscard]] static std::string CreateAndGetPresetDir();
+  [[nodiscard]] static std::string CreateAndGetPluginDir();
```
this is also not used anywhere - remove it.

reichlfl replied:
Done

dimitry- commented on OrbitCore/Path.h:
```diff
@@ -10,48 +10,47 @@
-  static std::string GetParentDirectory(std::string a_FullName);
-  static std::string JoinPath(const std::vector<std::string>& parts);
+  [[nodiscard]] static std::string GetFileName(const std::string& a_FullName);
+  [[nodiscard]] static std::string GetFileNameNoExt(const std::string& a_FullName);
```
unused - remove?

pierricgimmig replied:
if abseil or filesystem has the equivalent, I think we should remove them yes

reichlfl replied:
I think this is obsolete, not only because it's unused, but because `StripExtension(GetFileName())` provides the same. Removed.

dimitry- commented on OrbitCore/Path.h:
```diff
@@ -10,48 +10,47 @@
-  static std::string GetDirectory(const std::string& a_FullName);
-  static std::string GetParentDirectory(std::string a_FullName);
-  static std::string JoinPath(const std::vector<std::string>& parts);
+  [[nodiscard]] static std::string GetFileName(const std::string& a_FullName);
```
full_path

reichlfl replied:
I adjusted all those parameters also to hint at if they expect a file path, directory, or any path

dimitry- commented on OrbitCore/Path.h:
```diff
@@ -10,48 +10,47 @@
-  static std::string JoinPath(const std::vector<std::string>& parts);
+  [[nodiscard]] static std::string GetFileName(const std::string& a_FullName);
+  [[nodiscard]] static std::string GetFileNameNoExt(const std::string& a_FullName);
+  [[nodiscard]] static std::string StripExtension(const std::string& a_FullName);
```
and here... and below.

dimitry- commented on OrbitCore/Path.h:
```diff
@@ -10,48 +10,47 @@
+  [[nodiscard]] static std::string StripExtension(const std::string& a_FullName);
+  [[nodiscard]] static std::string GetExtension(const std::string& a_FullName);
+  [[nodiscard]] static std::string GetDirectory(const std::string& a_FullName);
+  [[nodiscard]] static std::string GetParentDirectory(std::string a_FullName);
```
not used

reichlfl replied:
Given that there's std::filesystem::path::parent_path() to provide this if we need it in the future, I removed it, thanks.

dimitry- commented on OrbitCore/Path.h:
```diff
@@ -10,48 +10,47 @@
   //  OrbitModule is removed.
-  static uint64_t FileSize(const std::string& a_File);
-  static bool DirExists(const std::string& a_Dir);
+  [[nodiscard]] static uint64_t FileSize(const std::string& a_File);
```
FileSize is not used anymore - remove it.

reichlfl replied:
And there's also `std::filesystem::file_size` in case we need it, so removed this.

dimitry- commented on OrbitCore/Path.h:
```diff
@@ -10,48 +10,47 @@
+  [[nodiscard]] static std::string GetFileName(const std::string& a_FullName);
+  [[nodiscard]] static std::string GetFileNameNoExt(const std::string& a_FullName);
+  [[nodiscard]] static std::string StripExtension(const std::string& a_FullName);
+  [[nodiscard]] static std::string GetExtension(const std::string& a_FullName);
```
and this is also unused.

reichlfl replied:
I'd argue we should keep this one. It's a method that fits perfectly in the scope of this utility class and neither abseil not std::filesystem provides it. I know we can always "get it back" from Git, but that requires someone to know it was there in the first place.

dimitry- replied:
I do not know what is the use-case for this one.

reichlfl replied:
It's now actually used in the new implementation of `StripExtension()` (bugfix for the unittest below), it makes that one quite straightforward and easy to follow...

antonrohr commented on OrbitCore/Path.h:
```diff
@@ -10,48 +10,47 @@
 
 #ifdef __linux__
-  static std::string GetHome();
+  [[nodiscard]] static std::string GetHome();
```
`GetHome()` is only used in `CreateAndGetOrbitAppDataDir()`. Maybe put the functionality of `GetHome()` into `CreateAndGetOrbitAppDataDir()` directly. Then we don't need this last ifdef in this header and only have ifdefs in the cpp file.

reichlfl replied:
Changed, GetHome() was a one-line anyways

antonrohr commented on OrbitCore/PathTest.cpp:
```diff
@@ -96,3 +96,171 @@ TEST(Path, JoinPathPartsAbsoluteRootBackslash) {
+             {"/", "/", "", ""}};
+#endif
+
+  for (size_t i = 0; i < tests.size(); ++i) {
```
Nit: 
```suggestion
  for (const auto& test: tests) {
```

reichlfl replied:
Done

antonrohr commented on OrbitCore/PathTest.cpp:
```diff
@@ -96,3 +96,171 @@ TEST(Path, JoinPathPartsAbsoluteRootBackslash) {
+  EXPECT_FALSE(Path::DirExists(Path::JoinPath({dir, "__invalid_subdir__"})));
+}
+
+TEST(Path, FileSize) {
```
Maybe add a test for FileSize of non-existing file. Currently the result is 0 in that case.

reichlfl replied:
Removed FileSize, so the test is obsolete

antonrohr commented on OrbitCore/PathTest.cpp:
```diff
@@ -96,3 +96,171 @@ TEST(Path, JoinPathPartsAbsoluteRootBackslash) {
+
+  for (auto fn : test_fns) {
+    std::string path = fn();
+    std::cout << "Testing existence of \"" + path + "\"\n";
```
Maybe use LOG instead of cout

antonrohr commented on OrbitCore/PathTest.cpp:
```diff
@@ -96,3 +96,171 @@ TEST(Path, JoinPathPartsAbsoluteRootBackslash) {
+
+// TODO: This would be more meaningful with absolute paths that are
+// created explicitely for the test
+TEST(Path, DirExists) {
```
Maybe add a test that checks the result of DirExist when called with a file instead of a directory

reichlfl replied:
Removed DirExists, so the test is obsolete

antonrohr commented on OrbitFramePointerValidator/FramePointerValidatorTest.cpp:
```diff
@@ -16,7 +16,7 @@ using orbit_grpc_protos::CodeBlock;
 
 TEST(FramePointerValidator, GetFpoFunctions) {
-  std::string executable_path = Path::GetExecutablePath();
+  std::string executable_path = Path::GetExecutableDir();
```
Nit:
```suggestion
  std::string executable_dit = Path::GetExecutableDir();
```

reichlfl replied:
Done

antonrohr commented on OrbitGl/ImGuiOrbit.cpp:
```diff
@@ -582,7 +582,7 @@ void Orbit_ImGui_RenderDrawLists(ImDrawData* draw_data) {
 
 ImFont* AddOrbitFont(float pixel_size) {
-  const auto exe_path = Path::GetExecutablePath();
+  const auto exe_path = Path::GetExecutableDir();
```
Nit:
```suggestion
  const auto exe_dir = Path::GetExecutableDir();
```

antonrohr commented on OrbitGl/TextRenderer.cpp:
```diff
@@ -55,7 +55,7 @@ void TextRenderer::Init() {
   m_Atlas = texture_atlas_new(atlasSize, atlasSize, 1);
 
-  const auto exePath = Path::GetExecutablePath();
+  const auto exePath = Path::GetExecutableDir();
```
Nit:
```suggestion
  const auto exe_dir = Path::GetExecutableDir();
```

reichlfl replied:
Done

antonrohr reviewed (commented):
Thank you a lot for this work. Especially the tests are really helpful!

I agree with @dimitry-'s comments an I have added a few more. 

Please also remove the `#pragma once` in Path.h. 

If you want to do even more work, you could change the whole thing to std::filesystem::path instead of std::string, this makes some small things nicer like concatenating paths with the / operator

pierricgimmig commented on OrbitCore/Path.h:
```diff
@@ -10,48 +10,47 @@
+  [[nodiscard]] static std::string GetDllName(bool a_Is64Bit);
+  [[nodiscard]] static std::string GetFileMappingFileName();
+  [[nodiscard]] static std::string GetSymbolsFileName();
+  [[nodiscard]] static std::string CreateAndGetCacheDir();
+  [[nodiscard]] static std::string CreateAndGetPresetDir();
+  [[nodiscard]] static std::string CreateAndGetPluginDir();
+  [[nodiscard]] static std::string CreateAndGetCaptureDir();
+  [[nodiscard]] static std::string CreateAndGetDumpDir();
+  [[nodiscard]] static std::string CreateAndGetOrbitAppDataDir();
```
Since the create operation will happen only once, should we rename to CreateOrGetDir instead of CreateANDGet? We use the GetOrCreateX pattern elsewhere in the code, GetOrCreateThreadTrack for example, maybe we should do the same here for uniformity.

reichlfl replied:
True, I changed this to be more consistent

dimitry- commented on OrbitCore/PathTest.cpp:
```diff
@@ -96,3 +96,171 @@ TEST(Path, JoinPathPartsAbsoluteRootBackslash) {
+  EXPECT_EQ(Path::StripExtension(input), "double.ext");
+  EXPECT_EQ(Path::StripExtension(Path::StripExtension(input)), "double");
+  EXPECT_EQ(Path::GetFileName(input), "double.ext.txt");
+  EXPECT_EQ(Path::GetFileNameNoExt(input), "double.ext");
```
add test for StripExtenstion("/some/dir.with.dots/filename_without_dots");

reichlfl replied:
Good point, and that actually uncovered a bug. Added the test and fixed this issue.

reichlfl commented:
@antonrohr @pierricgimmig @dimitry-  I think I addressed all comments now, let me know if I missed something. I generally do not think we should strictly remove every utility method that is not used if it makes sense to be in the class / namespace - but I removed most (I think all?) of the mentioned ones now since they were either super specific and won't be missed, or std::filesystem provides an alternative anyways. Thanks for the thorough feedback!

dimitry- commented on OrbitCore/PathTest.cpp:
```diff
@@ -96,3 +97,151 @@ TEST(Path, JoinPathPartsAbsoluteRootBackslash) {
+  EXPECT_EQ(Path::GetExtension(input), ".txt");
+  EXPECT_EQ(Path::StripExtension(input), "double.ext");
+  EXPECT_EQ(Path::StripExtension(Path::StripExtension(input)), "double");
+  EXPECT_EQ(Path::GetFileName(input), "double.ext.txt");
```
In general these should also work on linux-style directory separators on windows.. I do not think they should be platform-specific. It is not uncommon to parse a string we recieved via grpc from linux on windows. Or paths from .debug_info section which could be whatever os was using while compiling them.

reichlfl replied:
The test actually covers the first part (forward slash on windows) - only the first input is platform specific to check if both platforms handle their respective "native" formats for absolute paths correctly. All other tests are using linux-style directories for both platforms.

I changed one of the tests to cover a mixed-style path for both platforms. This exposed that the `StripExtension` isn't consistent with the other methods - all other methods that return a path convert it to contain only forward-slashes, so I changed `StripExtension` to do that as well, WDYT?

dimitry- commented on OrbitCore/PathTest.cpp:
```diff
@@ -96,3 +97,151 @@ TEST(Path, JoinPathPartsAbsoluteRootBackslash) {
+#else
+  std::string abs_input[] = {"/some_abs_path/file.txt", "/dir.with.dots/file_without_ext"};
+  std::string abs_stripped[] = {"/some_abs_path/file", "/dir.with.dots/file_without_ext"};
+#endif
```
also maybe remove this ifdefs and test for all 4 cases on both platforms?

reichlfl replied:
Right, actually that's the nicer solution. I've removed this, tests are now executed the same on both OS and should yield the same results for window-, unix- and mixed-style paths

antonrohr commented on OrbitCore/SymbolHelper.cpp:
```diff
@@ -52,7 +52,7 @@ std::vector<fs::path> ReadSymbolsFile() {
 
       const fs::path& dir = line;
-      if (fs::exists(dir)) {
+      if (std::filesystem::is_directory(dir)) {
```
Nit: 
```suggestion
      if (fs::is_directory(dir)) {
```
